File: src/main/java/com/huangyueran/spark/operator/Cogroup.java
package com.huangyueran.spark.operator;

import com.huangyueran.spark.utils.SparkUtils;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;
import scala.Tuple3;

import java.util.ArrayList;
import java.util.List;

/**
 * @category 对多个RDD中的KV元素，每个RDD中相同key中的元素分别聚合成一个集合。与reduceByKey不同的是针对两个RDD中相同的key的元素进行合并。
 * @author huangyueran
 * @time 2019-7-21 16:52:49
 */
public class Cogroup {

	public static void main(String[] args) {
		JavaSparkContext sc = SparkUtils.getLocalSparkContext(Cogroup.class);

		cogroup(sc);
	}

	private static void cogroup(JavaSparkContext sc) {
		List<Tuple2<Integer, String>> datas1 = new ArrayList<>();
		datas1.add(new Tuple2<>(1, "苹果"));
		datas1.add(new Tuple2<>(2, "梨"));
		datas1.add(new Tuple2<>(3, "香蕉"));
		datas1.add(new Tuple2<>(4, "石榴"));

		List<Tuple2<Integer, Integer>> datas2 = new ArrayList<>();
		datas2.add(new Tuple2<>(1, 7));
		datas2.add(new Tuple2<>(2, 3));
		datas2.add(new Tuple2<>(3, 8));
		datas2.add(new Tuple2<>(4, 3));

		List<Tuple2<Integer, String>> datas3 = new ArrayList<>();
		datas3.add(new Tuple2<>(1, "7"));
		datas3.add(new Tuple2<>(2, "3"));
		datas3.add(new Tuple2<>(3, "8"));
		datas3.add(new Tuple2<>(4, "3"));
		datas3.add(new Tuple2<>(4, "4"));
		datas3.add(new Tuple2<>(4, "5"));
		datas3.add(new Tuple2<>(4, "6"));

		/**
		 *   ===================================================== =========================
		 *   |    Cogroup: groups the elements in the same key in each RDD into a collection of KV elements in each RDD.                   |
		 *   |    Unlike reduceByKey, the elements of the same key are merged in the two RDD.                                                                   | 
		 *   ===============================================================================
		 */
		sc.parallelizePairs(datas1).cogroup(sc.parallelizePairs(datas2), sc.parallelizePairs(datas3)).foreach(
				new VoidFunction<Tuple2<Integer, Tuple3<Iterable<String>, Iterable<Integer>, Iterable<String>>>>() {
					@Override
					public void call(Tuple2<Integer, Tuple3<Iterable<String>, Iterable<Integer>, Iterable<String>>> t)
							throws Exception {
						System.out.println(t._1 + "==" + t._2);
					}
				});
	}

}

File: src/main/scala/com/hyr/spark/operator/Cogroup.scala
package com.hyr.spark.operator

import com.hyr.spark.utils.SparkUtils
import org.apache.spark.{SparkConf, SparkContext}

/** *****************************************************************************
  * @date 2019-08-08 08:39
  * @author: <a href=mailto:huangyr>黄跃然</a>
  * @Description: 对多个RDD中的KV元素，每个RDD中相同key中的元素分别聚合成一个集合。与reduceByKey不同的是针对两个RDD中相同的key的元素进行合并。
  ******************************************************************************/
object Cogroup {

  def cogroup(sparkContext: SparkContext): Unit = {
    val datas1 = List((1, "苹果"), (2, "梨"), (3, "香蕉"), (4, "石榴"))

    val datas2 = List((1, 7), (2, 3), (3, 8), (4, 3))

    val datas3 = List((1, "7"), (2, "3"), (3, "8"), (4, "3"), (4, "4"), (4, "5"), (4, "6"))

    val rdd1 = sparkContext.parallelize(datas1)
    val rdd2 = sparkContext.parallelize(datas2)
    val rdd3 = sparkContext.parallelize(datas3)

    val rdd = rdd1.cogroup(rdd2,rdd3)
    rdd.foreach(tuple => {
      println("key:" + tuple._1 + "\tvalue:" + tuple._2)
    })

  }

  def main(args: Array[String]): Unit = {
    val sparkContext = SparkUtils.getRemoteSparkContext(Cogroup.getClass)
    cogroup(sparkContext)
  }

}
